### ISSUE TITLE
[GSoC] Logarithm bases `\lg` and `\log`

### ISSUE THREAD

**Issue by wermos** [sympy/sympy#25420]
## Background

The ANTLR LaTeX Parser currently parsers `\log` as base `e`, and `\lg` as base 10. While writing the Lark Parser, I wondered how useful this is, because I've only seen `\log` to denote log base 10, and `\lg` as the binary logarithm (i.e., log base 2) in computational complexity-related things.

ANTLR cites ISO 80000-2:2019 as its reference for why `\lg` uses base 10.

When I asked my GSoC mentor @Upabjojr about this, he said that he too believes that `\log` should use base `e`.

## Lark Parser

As of right now, I've preserved the ANTLR choice of bases in the Lark rewrite, but once again, I wanted to get the community's thoughts on what base the logarithm should use for both `\log` and `\lg`.

**Comment by asmeurer** [comment]
log should definitely default to base e

For lg, we should try to see what it is actually used as. In general, we should try to match common usage, rather than ISO standards.

**Comment by wermos** [comment]
> log should definitely default to base e

I agree. The reasoning given by @eagleoflqj of matching the default of the LaTeX printer, in addition to the reasons, is good enough to me.

> For lg, we should try to see what it is actually used as. In general, we should try to match common usage, rather than ISO standards.

I agree with this too. Do you have any thoughts on how we might go about this?

**Comment by asmeurer** [comment]
Maybe we can try to find uses on arxiv and google scholar. We could at least get an idea of what fields actually use that notation and how they define it.

**Comment by wermos** [comment]
> Maybe we can try to find uses on arxiv and google scholar. We could at least get an idea of what fields actually use that notation and how they define it.

I used [this](https://arxiv.org/search/advanced?advanced=&terms-0-operator=AND&terms-0-term=%24%5Clg%24&terms-0-field=all&classification-physics_archives=all&classification-include_cross_list=include&date-filter_by=all_dates&date-year=&date-from_date=&date-to_date=&date-date_type=submitted_date&abstracts=show&size=50&order=-announced_date_first) search query on ArXiv, and the search returned two CS papers:

1. [Indexing Graph Search Trees and Applications](https://arxiv.org/abs/1906.07871)
2. [Biconnectivity, $st$-numbering and other applications of DFS using $O(n)$ bits](https://arxiv.org/abs/1606.08645)

Both papers say
> \footnote{We use lg to denote logarithm to the base 2.}

Admittedly, both papers share the same person as first author, so I'll try to look for more evidence for or against $\lg$ as the base 2 logarithm.

**Comment by Upabjojr** [comment]
I have always thought of `\lg` as base 2 logarithm.

I would suggest to create a node for `\lg` in the Lark parser and then deal with the transformation into the proper logarithm base in the transformer part of the parser. The transformer could have some transformation rules. For example:

In Lark:
```
LG_NODE: "\lg" expression_mul
```
In the transformer:
```python
 class TreeToSympy(_lark.Transformer):
    def __init__(self, transformations={"LG_NODE": sympy.log2}):
        self.transformations = transformations

    def LG_NODE(self, token):
        return self.transformations["LG_NODE"](token)
```
in this way, if a user doesn't like the default interpretation of "\lg", he can pass an alternative dictionary of transformations to the parser.

**Comment by wermos** [comment]
> in this way, if a user doesn't like the default interpretation of "\lg", he can pass an alternative dictionary of transformations to the parser.

This is a good idea, but I think this ties into the more general discussion we should have, which is: What should we leave as user-configurable, and how should we allow for user-configuration in the Lark parser?

I'll open an issue about that soon, so that we can start discussing how to this issue, along with a few other related issues.

**Comment by asmeurer** [comment]
User configurability is indeed something that I'm hopeful we can get a lot of with the Lark parser. Even so, we should have good defaults so that the majority of users don't have to worry about it.

**Comment by sylee957** [comment]
I just expect that user configurable functions to have type annotations

**Comment by asmeurer** [comment]
@sylee957 can you clarify what you mean? How do you imagine the user configurability for the latex parser to look like? 

Personally I see two types of configurability. One is changing the meaning of an existing parsed node, like changing `\lg` to parse as `log(x)` instead of `log(x, 2)`. That could probably just be done by passing a dictionary to `parse_latex`. The second would be for the user to extend the grammar itself to be able to parse things that aren't otherwise parsable. I'm not sure yet what that should look like, and it probably depends on the implementation details of how Lark works. Maybe @Upabjojr has ideas here.

**Comment by asmeurer** [comment]
But at any rate, I'm not clear how type safety factors in.

**Comment by sylee957** [comment]
I expect users to know what the *** is input and output of `LG_NODE` it is. 
The best way to do this is by type system, and the worst is by expecting the users to read the source code, even though it is open source.

**Comment by sylee957** [comment]
And i also recommend to factor out your own grammar in modules, 

https://lark-parser.readthedocs.io/en/stable/grammar.html#import

with its module system before getting an insight about how user-configurable grammar itself should look like. 

I can share some tricks for more tricky refactory, like factoring out recursive grammar (like arithmetic), which I had used to do before at work.

**Comment by asmeurer** [comment]
@sylee957 thanks for the tips. If you want to be more involved in the project you can join the Matrix chat https://matrix.to/#/#sympy-gsoc-2023-latex-parser:matrix.org (we can also add you to the weekly video calls if you are interested). Having someone who's worked with this stuff before would indeed be helpful.

**Comment by asmeurer** [comment]
I'm a little unsure about typing. The only way to know what type an expression node will really support is to create it, because SymPy nodes themselves are not really typed well enough to check that. I think we'd want to have SymPy itself be typed before we can heavily type the parser. And in LaTeX, you can in principle create anything, even nonsense that doesn't make any sense type-wise, but there'd be no way to know that it is nonsense until you try to create the SymPy expression. If anything, I'd think the parser nodes should not be typed, and just represent structurally whatever the user input, and let the actual typing and representation of semantic mathematical meaning come after translating it to SymPy.

**Comment by sylee957** [comment]
It can be possible not to type, however, in that regards, you should make the options to be very simple, not more than boolean flags, univariate function callbacks or such.

**Comment by asmeurer** [comment]
I generally agree that we should try not to have a hundred flags to `parse_latex` (unlike `latex()`). There should be a lower level API that people should use if they need to customize things, similar to how things work with the printers.

**Comment by sylee957** [comment]
> The only way to know what type an expression node will really support is to create it, because SymPy nodes themselves are not really typed well enough to check that.

I don’t think that LaTeX parser should be tied heavily to the SymPy.
It can be one-shot code generator of math, numpy, or other computer algebra systems as well if you switch the transformer, if it is designed well.

However, back to the point if we need options, I’m also kind of skeptical that people can just copy paste the source code than depending on options,
(Because sympy is anyway open source), as long as we stay simple

For example, I have used to reference how the old Antlr parser was designed, but had not used that as library because I had lots of things to configure, like handling some other educational notations or parsing Vietnamese math, which you may not have an easy idea to abstract.

**Comment by asmeurer** [comment]
> I don’t think that LaTeX parser should be tied heavily to the SymPy.

I agree, that's why I think it should try to parse to an AST that is purely based on the structure of the input latex, without any mathematical types. That AST should then be converted to SymPy.

**Comment by bjodah** [comment]
In chemistry `log` is base 10 (horrible, I know). I thought it was generally agreed upon that the two letter log abbreviations maps to bases as: `{'lb': 2, 'ln': e, 'lg': 10}`. But I guess the only truly unambiguous form is $\mathrm{log}_x$.

**Comment by wermos** [comment]
> And to add to the confusion: I think my Texas TI-83 even coded the button LOG to base 10, and LN to base $e$. ¯\_(ツ)_/¯

Yes, I think the `log = base 10` and `ln = base e` convention is widely used in engineering as well. The calculator I have coded the buttons the same way. AFAIK, `log = base e` is a largely mathematician convention.
